MetadataExtractor/IO/RandomAccessStream.cs

```diff
@@ -0,0 +1,578 @@
+
+            // This bypasses a lot of checks particularly when the input was a byte[]
+            // TODO: good spot to try Span<T>
+            if (p_isStreamFinished && p_chunks.Count == 1)
+            {
+                Array.Copy(p_chunks[0], (int)index, buffer, 0, count);
+                return count;
+            }
+
+            var remaining = count;                      // how many bytes are requested
+            var fromOffset = (int)index;
+            var toIndex = offset > 0 ? offset : 0;
+            while (remaining != 0)
+            {
+                var fromChunkIndex = fromOffset / p_chunkLength;     // chunk integer key
+                var fromInnerIndex = fromOffset % p_chunkLength;     // index inside the chunk to start reading
+                var length = Math.Min(remaining, p_chunkLength - fromInnerIndex);
+                var chunk = p_chunks[fromChunkIndex];
+                Array.Copy(chunk, fromInnerIndex, buffer, toIndex, length);
+                remaining -= length;
+                fromOffset += length;
+                toIndex += length;
+            }
+
+            return toIndex - offset;
+        }
+
+        /// <summary>Returns an unsigned byte at an index in the sequence.</summary>
+        /// <returns>the 8 bit int value, between 0 and 255</returns>
+        /// <param name="index">position within the data buffer to read byte</param>
+        /// <param name="isSequential">flag indicating if caller is using sequential access</param>
+        /// <exception cref="BufferBoundsException"/>
+        /// <exception cref="IOException"/>
+        public byte GetByte(long index, bool isSequential)
+        {
+            return GetByte(index, isSequential, true);
+        }
+
+        /// <summary>Returns an unsigned byte at an index in the sequence.</summary>
+        /// <returns>the 8 bit int value, between 0 and 255</returns>
+        /// <param name="index">position within the data buffer to read byte</param>
+        /// <param name="isSequential">flag indicating if caller is using sequential access</param>
+        /// <param name="validateIndex">allows for skipping validation if already done by the caller</param>
+        /// <exception cref="BufferBoundsException"/>
+        /// <exception cref="IOException"/>
+        private byte GetByte(long index, bool isSequential, bool validateIndex)
+        {
+            if(validateIndex)
+                ValidateIndex(index, 1, isSequential);
+
+            // This bypasses a lot of checks particularly when the input was a byte[]
+            if (p_isStreamFinished && p_chunks.Count == 1)
+                return p_chunks[0][index];
+
+            var chunkIndex = index / p_chunkLength;
+            var innerIndex = index % p_chunkLength;
+
+            if (p_chunks.ContainsKey(chunkIndex))
+                return p_chunks[chunkIndex][innerIndex];
+            else
+                return unchecked((byte)-1);
```

Comment:
This will be `255` which is indistinguishable by the caller from a real value. Should this just throw?

Reply:
There were some specific metadata cases where GetByte was expected to return this value... but now I can't remember how that went. This one might have to be tested against the metadata repo again to figure out -- but that's down the line.

It might have had something to do with replicating the Stream.ReadByte() return value and behavior. When it can't read any more data, it returns -1. That isn't an error condition to the Stream, but SequentialStreamReader throws an IOException on that value. I could maybe try to throw the same exception and see if it works.

Reply:
I went ahead and changed it to throw for now. Will need to see how it works in more extensive testing.

Reply:
I think `GetByte` used to return `int`, which covers the domain of `byte` (which is unsigned) while allowing the special `-1` value to indicate something else. Casting `-1` to `byte` here pushed the value into the range of `byte` which means the caller can no longer tell that it was meant to be special. If this is going to return `byte`, then the exception is that special value.